### PR TITLE
Provide context parsing in the `mkMessage` function

### PR DESCRIPTION
**Pull request by Oleksandr-Zhabenko**
Provided context parsing in the mkMessage function. Resolves [issue  #282: mkMessage should support context parsing](https://github.com/yesodweb/shakespeare/issues/282)

CI passed successfully. See the link below:
[https://github.com/Oleksandr-Zhabenko/shakespeare/actions/runs/10403325875](https://github.com/Oleksandr-Zhabenko/shakespeare/actions/runs/10403325875)

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -1,5 +1,7 @@
 name: Tests
 
+# Is taken from the https://github.com/yesodweb/yesod/blob/master/.github/workflows/tests.yml template
+
```
Comment (jezen) [inline .github/workflows/tests.yml:4]: This comment probably isn't necessary.

File: Text/Shakespeare/I18N.hs [hunk 2]
```diff
@@ -62,22 +63,22 @@ module Text.Shakespeare.I18N
     ) where
 
 import Language.Haskell.TH.Syntax hiding (makeRelativeToProject)
-import Control.Applicative ((<$>))
+--import Control.Applicative ((<$>))
```
Comment (jezen) [inline Text/Shakespeare/I18N.hs:66]: If this import isn't necessary, it's better just to delete the line.

File: Text/Shakespeare/I18N.hs [hunk 3]
```diff
@@ -62,22 +63,22 @@ module Text.Shakespeare.I18N
...
 import Data.Text.Encoding (decodeUtf8)
-import Data.Char (isSpace, toLower, toUpper)
+import Data.Char (isSpace, toLower, toUpper, isLower)
 import Data.Ord (comparing)
 import Text.Shakespeare.Base (Deref (..), Ident (..), parseHash, derefToExp)
-import Text.ParserCombinators.Parsec (parse, many, eof, many1, noneOf, (<|>))
+import Text.ParserCombinators.Parsec (parse, many, eof, many1, noneOf, (<|>), string, spaces, char, option, alphaNum, sepBy1, try)
 import Control.Arrow ((***))
-import Data.Monoid (mempty, mappend)
+--import Data.Monoid (mempty, mappend)
```
Comment (jezen) [inline Text/Shakespeare/I18N.hs:81]: We can delete this line also.

File: Text/Shakespeare/I18N.hs [hunk 4]
```diff
@@ -165,20 +166,63 @@ mkMessageCommon genType prefix postfix master dt rawFolder lang = do
...
+                                then VarT $ mkName t
+                                else ConT $ mkName t
+
+                 isTvar :: String -> Bool  -- Is taken from the https://hackage.haskell.org/package/yesod-core-1.6.26.0/docs/src/Yesod.Routes.Parse.html#isTvar
+                 isTvar (h:_) = isLower h
+                 isTvar _     = False
+
+                 parseName = do
+                      cxt' <- option [] parseContext
+ --                     name' <- parseWord    -- This one is no longer needed as a separate argument
```
Comment (jezen) [inline Text/Shakespeare/I18N.hs:200]: Let's just keep the code that we need 🙂

Comment (Oleksandr-Zhabenko) [reply Text/Shakespeare/I18N.hs:200]: Ok. I will clean up the source files.

**Review by jezen — CHANGES_REQUESTED** [review]
Thanks for tidying up. Now we just need to demonstrate that this is not a breaking change, and that it produces correct code for the scenario documented in the associated issue (and we should squash before merging).

File: ChangeLog.md [hunk 5]
```diff
@@ -1,9 +1,5 @@
 # ChangeLog for shakespeare
 
-### 2.1.1
-
-* Add support for `TypeApplications` inside Shakespeare quasiquotes
-
```
Comment (jezen) [inline ChangeLog.md:6]: Is there a good reason why this changelog entry was removed? This looks like a mistake.

File: Text/Shakespeare/I18N.hs [hunk 6]
```diff
@@ -255,13 +301,13 @@ conP name = ConP name []
 conP = ConP
 #endif
 
-toCon :: String -> SDef -> Con
+toCon :: String -> SDef -> Con
```
Comment (jezen) [inline Text/Shakespeare/I18N.hs:304]: Let's not introduce trailing whitespace.

File: shakespeare.cabal [hunk 7]
```diff
@@ -1,5 +1,5 @@
 name:            shakespeare
-version:         2.1.1
+version:         2.1.0
```
Comment (jezen) [inline shakespeare.cabal:2]: I would expect this number to go up, not down.

File: test/Text/Shakespeare/BaseSpec.hs [hunk 8]
```diff
@@ -33,31 +33,6 @@ spec = do
...
-      Right
-        (DerefBranch
-          (DerefIdent (Ident "x"))
-          (DerefType "Maybe String"))
-  it "parseDeref parse single @ as operator" $ do
-    runParser parseDeref () "" "x @ y" `shouldBe`
-      Right
-        (DerefBranch
-          (DerefBranch (DerefIdent (Ident "@")) (DerefIdent (Ident "x")))
-          (DerefIdent (Ident "y")))
```
Comment (jezen) [inline test/Text/Shakespeare/BaseSpec.hs:59]: Why were these tests removed?

File: Text/Shakespeare/I18N.hs [hunk 9]
```diff
@@ -165,22 +164,69 @@ mkMessageCommon genType prefix postfix master dt rawFolder lang = do
...
-            []
-            (ConT ''RenderMessage `AppT` (ConT $ mkName master) `AppT` ConT mname)
+            cxt  -- Here the parsed context should be added, otherwise []
+            (ConT ''RenderMessage `AppT` (if ' ' `elem` master' then let (ts, us) = break (== ' ') . filter (\x -> x /= '(' && x /= ')') $ master' in let us1 = filter (/= ' ') us in ParensT (ConT (mkName ts) `AppT` VarT (mkName us1))  else ConT $ mkName master') `AppT` ConT mname)
             [ FunD (mkName "renderMessage") $ c1 ++ c2 ++ [c3]
             ]
         ]
-
-toClauses :: String -> String -> (Lang, [Def]) -> Q [Clause]
+           where (dt1, cxt0) = case (parse parseName "" dt) of
```
Comment (jezen) [inline Text/Shakespeare/I18N.hs:181]: Let's improve the formatting of all the new parsing code here.

File: test/Text/Shakespeare/BaseSpec.hs (hunk 8 above)
Comment (Oleksandr-Zhabenko) [reply test/Text/Shakespeare/BaseSpec.hs:59]: I do not know. Should I add these lines into the BaseSpec.hs?

File: ChangeLog.md (hunk 5 above)
Comment (Oleksandr-Zhabenko) [reply ChangeLog.md:6]: On Hackage the version 2.1.1 is the last one and is already deprecated.
[https://hackage.haskell.org/package/shakespeare-2.1.1](https://hackage.haskell.org/package/shakespeare-2.1.1)

File: shakespeare.cabal (hunk 7 above)
Comment (Oleksandr-Zhabenko) [reply shakespeare.cabal:2]: I think it can be changed to 2.1.2

File: ChangeLog.md (hunk 5 above)
Comment (jezen) [reply ChangeLog.md:6]: Ok, but it shouldn't be removed even if it's deprecated. The deprecated version should continue to exist in the version history.

File: test/Text/Shakespeare/BaseSpec.hs (hunk 8 above)
Comment (jezen) [reply test/Text/Shakespeare/BaseSpec.hs:59]: Looks like a minor git accident then 🙂 You can just revert this change.

**Pull request closed without merging**